### PR TITLE
feat(perf): add microbenchmark suite for cat_ranges

### PR DESCRIPTION
**Pull request by yuxin00j**
## Overview

This PR adds a comprehensive microbenchmark suite for `cat_ranges` under `gcsfs/tests/perf/microbenchmarks/cat_ranges`. It benchmarks multi-range read throughput and latency across Regional, Zonal (Rapid), and Hierarchical Namespace (HNS) buckets under various access patterns and parameter combinations.

## Summary of Changes

1. **Benchmark Suite (`gcsfs/tests/perf/microbenchmarks/cat_ranges/`)**:
   - `parameters.py`: Defines `CatRangesBenchmarkParameters` capturing range counts, range selection patterns, chunk sizes, `max_gap`, and `batch_size`.
   - `configs.py`: Implements `CatRangesConfigurator` to generate test matrices from YAML configurations.
   - `configs.yaml`: Defines baseline scenarios covering:
     - `cat_ranges_seq`: Sequential adjacent ranges across a file.
     - `cat_ranges_rand`: Uniform random ranges across a file.
     - `cat_ranges_mixed_sizes`: Variable chunk sizes (64 KB, 1 MB, 4 MB, 16 MB) within the same call.
     - `cat_ranges_multi_file`: Multi-file range requests distributed round-robin across 5 files.
     - `cat_ranges_batch_size_sweep`: Concurrency sweep evaluating batch sizes 16, 64, and 128.
     - `cat_ranges_large_file_large_ranges`: Scale scenario requesting 1,000 mixed ranges across a 10 GB file with batch size 64.
   - `test_cat_ranges.py`: Implements `test_cat_ranges_single_threaded` benchmark runner using `pytest-benchmark` and pre-generated range offsets to avoid measuring setup overhead.

2. **Fixtures and Metrics Reporting (`gcsfs/tests/perf/microbenchmarks/conftest.py`)**:
   - Added `gcsfs_benchmark_cat_ranges` fixture to manage lifecycle of temporary test files across bucket types.
   - Enhanced `publish_benchmark_extra_info` to record `num_ranges`, `max_gap`, and `batch_size` in the benchmark report and output CSVs.

3. **Test Framework Registration (`gcsfs/tests/conftest.py`)**:
   - Registered `cat_ranges` in `benchmark_subdirs` to ensure test collection when passing `--run-benchmarks`.

4. **Unit Test Coverage (`gcsfs/tests/perf/microbenchmarks/test_configs.py`)**:
   - Added `test_cat_ranges_configurator` unit test.
   - Registered `get_cat_ranges_benchmark_cases()` in `test_validate_actual_yaml_configs` to validate YAML integrity in CI.

### CODE REVIEW

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        # 3. Determine start and end offsets within file bounds
+        max_offset = max(0, file_size_bytes - range_size)
+        if pattern == "seq":
+            start = current_seq_offset % max(1, max_offset) if max_offset > 0 else 0
+            current_seq_offset += range_size
+        else:
+            start = rng.randint(0, max_offset) if max_offset > 0 else 0
+
+        starts.append(start)
+        ends.append(start + range_size)
```
[the remark this answers is not in the document]
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:82]: Done. Stopped sequential range generation when `current_seq_offset > max_offset`, added validation ensuring `range_size <= file_size_bytes`, and added unit tests.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/configs.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        pattern = scenario.get("pattern", "rand")
+        scenario_files = scenario.get("files", 1)
+
+        raw_chunk_sizes = scenario.get(
+            "chunk_sizes_mb", common_config.get("chunk_sizes_mb", [[1]])
+        )
+        if raw_chunk_sizes and not isinstance(raw_chunk_sizes[0], list):
+            chunk_size_combinations = [raw_chunk_sizes]
+        else:
+            chunk_size_combinations = raw_chunk_sizes
```
[the remark this answers is not in the document]
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/configs.py:33]: Done. Added fallback default to `[[1]]` when `raw_chunk_sizes` is empty or falsy.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/configs.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        batch_sizes = scenario.get(
+            "batch_sizes", common_config.get("batch_sizes", [None])
+        )
+        max_gaps = scenario.get("max_gaps", common_config.get("max_gaps", [None]))
```
[the remark this answers is not in the document]
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/configs.py]: Done. Added `or [None]` fallbacks for `batch_sizes` and `max_gaps`.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Yonghui-Lee) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:82]: If we stop when current_seq_offset > max_offset, it may skew benchmark results because a truncated run will look faster than it actually is.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    rng = random.Random(seed)
+    paths = []
+    starts = []
+    ends = []
+
+    current_seq_offset = 0
+    for i in range(num_ranges):
+        # 1. Path allocation (round-robin across available files)
+        path = file_paths[i % len(file_paths)]
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:55]: if `pattern == "seq"` is used with `files > 1`:

- file_0 gets [0, 1MB].
- file_1 gets [1MB, 2MB].
- file_2 gets [2MB, 3MB].
- file_0 gets [3MB, 4MB]. 

None of the files experience sequential range reads.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            start = rng.randint(0, max_offset) if max_offset > 0 else 0
+
+        paths.append(path)
+        starts.append(start)
+        ends.append(start + range_size)
+
+    return paths, starts, ends
+
+
+def _cat_ranges_op(gcs, paths, starts, ends, max_gap=None, batch_size=None):
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:87]: In fsspec, `cat_ranges(..., on_error="return")` is the default. Because _cat_ranges_op does not pass `on_error="raise"` or inspect the return value, partial or total failures will silently succeed and record falsely low latencies in the benchmark. We can pass `on_error="raise"` to cat_ranges.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/parameters.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+@dataclass
+class CatRangesBenchmarkParameters(IOBenchmarkParameters):
+    """
+    Defines the parameters for a cat_ranges benchmark test case.
+    """
+
+    # Number of byte ranges to fetch in a single cat_ranges call
+    num_ranges: int
+
+    # Range selection pattern: "seq", "rand", or "mixed"
```
Comment (Yonghui-Lee) [inline gcsfs/tests/perf/microbenchmarks/cat_ranges/parameters.py]: We don't support mixed pattern in `_generate_ranges`?

**Review by Yonghui-Lee — COMMENTED** [review]
Throughput calculation in run.py is wrong for cat_ranges. For cat_ranges, the actual data transferred is not `file_size * files`.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py (hunk 5 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:87]: Done.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py (hunk 4 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py:55]: Fixed. Maintains independent offsets per file.

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/parameters.py (hunk 6 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/parameters.py]: Fixed. Removed mixed.

File: cloudbuild/benchmarks/benchmarks_schema.json [hunk 7]
```diff
@@ -35,6 +35,9 @@
       {"name": "seq_probability", "type": "STRING"},
       {"name": "engine", "type": "STRING"},
       {"name": "method", "type": "STRING"},
+      {"name": "batch_size", "type": "INTEGER"},
+      {"name": "max_gap", "type": "INTEGER"},
+      {"name": "num_ranges", "type": "INTEGER"},
```
Comment (zhixiangli) [inline cloudbuild/benchmarks/benchmarks_schema.json:40]: Do we need total_bytes as well?

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        if isinstance(chunk_sizes_bytes, list)
+        else [chunk_sizes_bytes]
+    )
+    if max(chunk_sizes) > file_size_bytes:
+        raise ValueError(
+            f"Range size {max(chunk_sizes)} bytes exceeds file size {file_size_bytes} bytes."
+        )
+    if (
+        pattern == "seq"
+        and (num_ranges / len(file_paths)) * max(chunk_sizes) > file_size_bytes
```
Comment (zhixiangli) [inline gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py]: nit: How about moving the validation to line 80? There may be edge cases that the floor division of `num_ranges / len(file_paths)` and `max(chunk_sizes)` does not handle properly.

File: cloudbuild/benchmarks/benchmarks_schema.json (hunk 7 above)
Comment (yuxin00j) [reply cloudbuild/benchmarks/benchmarks_schema.json:40]: Added

File: gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py (hunk 8 above)
Comment (yuxin00j) [reply gcsfs/tests/perf/microbenchmarks/cat_ranges/test_cat_ranges.py]: Done.

**Review by zhixiangli — APPROVED** [review]
